File: Matlab/CCD_2D.m
function [ output_angles, output_positions ] = CCD_2D( current_angles,...
    current_positions, constraints, goal, threashold, max_iterations )
%CCD_2D uses cyclic coordinate descent to calculate joint rotations for an
%inverse kinematic solution to a revolute joint 2d robot
%
%   output_angles: 1 x n-1 array of calculated joint angles
%
%   output_positions: 2 x n array of calculated joint positions [x1, x2, ...,
%   xn; y1, y2, ..., yn];
%
%   current_angles: 1 x n-1 array of input joint angles
%
%   current_positions: 2 x n array of current joint positions [x1, x2, ...,
%   xn; y1, y2, ..., yn];
%
%   constraints: 2 x n-1 array of max/min joint angles
%
%   goal: 2 x 1 array of desired end effector position
%
%   threashold: single value, tolerance for how close the end effector
%   needs to be to the goal before returning
%
%   max_iterations: maximum CCD iterations before returning

%%  Inputs

num_of_link = length(current_positions)-1; %number of links
%   Retrieve initial joint X locations
xdata = current_positions(1,:);
%   Retrieve initial joint Y locations
ydata = current_positions(2,:);
%   Retrieve initial joint angles
angledata = current_angles;
jointAngle = angledata;
output_angles = jointAngle;

%%  Initialization of data

error = dist([xdata(num_of_link+1) ydata(num_of_link+1)], goal);
num_iterations = 0;
while (error > threashold && num_iterations < max_iterations)
    active_joint = num_of_link + 1;
    while (active_joint > 1) 
        %% CCD Algorthm
        % end effector position
        pe = [xdata(num_of_link+1); ydata(num_of_link+1)];
        % current joint position
        pc = [xdata(active_joint-1); ydata(active_joint-1)];
        % vector from end effector to current joint
        a = (pe - pc)/norm(pe-pc);
        % vector from current joint to target
        b = (goal - pc)/norm(goal-pc);
        % rotation angle for current joint to minimize b
        theta = acosd(dot(a, b));
        % check for rotation angle
        direction = cross([a(1) a(2) 0],[b(1) b(2) 0]);
        if direction(3) < 0
            theta = -theta;
        end
        
        % Joint Constraints
        if (theta > 60)
            theta = 60;
        elseif (theta < -60)
            theta = -60;
        end

        angledata(active_joint) = theta;
        jointAngle(active_joint) = theta;

        
        %% Perform Rotation
        
        %   For first joint
        i = active_joint;
        %   Calculate rotation matrix
        R = [cosd(angledata(i)) -sind(angledata(i)); sind(angledata(i)) cosd(angledata(i))]; % rotation matrix
        %   Rotate outer point around joint
        temp = R * ([xdata(i); ydata(i)] - [xdata(i-1); ydata(i-1)]) + [xdata(i-1); ydata(i-1)];
        xdata(i) = temp(1);
        ydata(i) = temp(2);
        angledata(i) = angledata(i) + angledata(i-1);
        
        %   For next joint(s)
        i = active_joint+1;
        while (i <= num_of_link+1)
            %   Rotate each joint around active joint
            temp = R * ([xdata(i); ydata(i)] - [xdata(active_joint-1); ydata(active_joint-1)]) + [xdata(active_joint-1); ydata(active_joint-1)];
            xdata(i) = temp(1);
            ydata(i) = temp(2);
            angledata(i) = angledata(i) + angledata(i-1);
            i = i+1;
        end
        
        % advance iteration
        active_joint = active_joint - 1;

    end
    % calculate error
    error = dist([xdata(num_of_link+1) ydata(num_of_link+1)], goal);
    % update output angles with current iteration deltas
    output_angles = output_angles + jointAngle;
    % add to iteration count
    num_iterations = num_iterations + 1;
end

output_positions = [xdata;ydata];

end

File: Python/CCD_2D.py
import math
import numpy as np
from numpy import pi
import matplotlib.pyplot as plt
import pdb

def CCD_2D( x_data, y_data, angle_data, constraints, goal, threashold, max_iterations):
    "This does stuff"
    
    n_link = x_data.size-1

    # reshape all matrices
    x_data = x_data.reshape(n_link+1,1)
    y_data = y_data.reshape(n_link+1,1)
    angle_data = angle_data.reshape(n_link+1,1)
    goal = goal.reshape(2,1)

    # cast all matrices to float
    x_data = x_data.astype(float)

    jointAngle = angle_data
    output_angles = angle_data
    # Initialization of Data
    # Set up plot (DEBUG)
    plt.axis([-1, 10, -1, 7])
    plt.ion()
    plt.show()
    # plot data (DEBUG)
    line, = plt.plot(x_data,y_data,'r-')
    plt.plot(goal[0],goal[1],'b*')
    plt.draw()
    plt.pause(0.001)
    raw_input("Press [Enter] to continue")
    
    error = np.linalg.norm(np.array([(x_data[n_link]),(y_data[n_link])])-goal)

    #pdb.set_trace()

    n_iterations = 0

    while (error > threashold and n_iterations < max_iterations):
        active_joint = n_link + 1
        while active_joint > 1:
            # CCD Algorithm
            # End Effector Position
            pe = np.array([x_data[n_link],y_data[n_link]])
            # Current Joint Position
            pc = np.array([x_data[active_joint-2],y_data[active_joint-2]])
            # vector from end effector to current joint
            a = (pe-pc)/np.linalg.norm(pe-pc)
            a = a.reshape(1,2)
            # vector from current joint to target
            b = (goal-pc)/np.linalg.norm(goal-pc)
            # rotation angle for current joint to minimize b
            theta = math.acos(np.dot(a,b))
            print("a")
            print(a)
            print("b")
            print(b)
            print("Theta")
            print(math.degrees(theta))
            #pdb.set_trace() # DEBUG
            # check for rotation direction
            b = b.reshape(1,2)
            direction = np.cross(np.array([a[0,0],a[0,1],0]),np.array([b[0,0],b[0,1],0]))
            if direction[2] < 0:
                theta = -theta
            # joint constraints
            if theta > (pi/3):
                theta = (pi/3)
            elif theta < -(pi/3):
                theta = -(pi/3)
            angle_data[active_joint-1]=theta
            jointAngle[active_joint-1]=theta

            # Perform Rotation

            # for first joint
            i = active_joint
            # calculate rotation matrix
            R = np.array([(math.cos(angle_data[i-1]),-math.sin(angle_data[i-1])),(math.sin(angle_data[i-1]),math.cos(angle_data[i-1]))])

            pdb.set_trace() # DEBUG
            while i<=n_link+1:
                #pdb.set_trace() # DEBUG
                # rotate each joint around active joint
                temp = R.dot(np.array([x_data[i-1]-x_data[active_joint-2],y_data[i-1]-y_data[active_joint-2]])) + np.array([x_data[active_joint-2],y_data[active_joint-2]])
                x_data[i-1] = temp[0,0]
                y_data[i-1] = temp[1,0]
                angle_data[i-1] = angle_data[i-1] + angle_data[i-2]
                print("temp")
                print(temp)
                print("x_data")
                print(x_data)
                print("y_data")
                print(y_data)
                print("angle_data")
                print(angle_data)

                # plot data (DEBUG)
                line.set_xdata(x_data)
                line.set_ydata(y_data)
                plt.draw()
                plt.pause(0.001)
                raw_input("Press [Enter] to continue")

                #pdb.set_trace() # DEBUG
                i = i + 1
            # advance iteration
            active_joint = active_joint - 1
        # calculate error
        error = np.linalg.norm(np.array([(x_data[n_link]),(y_data[n_link])])-goal)
        # update output angles with current iteration deltas
        output_angles = output_angles + jointAngle
        # add to iteration count
        n_iterations = n_iterations + 1
    return (x_data,y_data,angle_data)

